### DISCUSSION TITLE
Support for MicroPython

### DISCUSSION THREAD

**Discussion by rgov** [foxglove/mcap#1091] [General]
I am interested in having a microcontroller record data directly to MCAP format from [MicroPython](https://micropython.org). I gave it a try today using the UNIX port.

First up we need a [package.json](https://docs.micropython.org/en/latest/reference/packages.html#writing-publishing-packages) file. (Since this file isn't in the GitHub repo, I hosted it locally then used `micropython -m mip install http://localhost:8000/package.json`.)

```json
{
    "urls": [
        ["mcap/__init__.py", "github:foxglove/mcap/python/mcap/mcap/__init__.py"],
        ["mcap/_chunk_builder.py", "github:foxglove/mcap/python/mcap/mcap/_chunk_builder.py"],
        ["mcap/_message_queue.py", "github:foxglove/mcap/python/mcap/mcap/_message_queue.py"],
        ["mcap/data_stream.py", "github:foxglove/mcap/python/mcap/mcap/data_stream.py"],
        ["mcap/decoder.py", "github:foxglove/mcap/python/mcap/mcap/decoder.py"],
        ["mcap/exceptions.py", "github:foxglove/mcap/python/mcap/mcap/exceptions.py"],
        ["mcap/opcode.py", "github:foxglove/mcap/python/mcap/mcap/opcode.py"],
        ["mcap/reader.py", "github:foxglove/mcap/python/mcap/mcap/reader.py"],
        ["mcap/records.py", "github:foxglove/mcap/python/mcap/mcap/records.py"],
        ["mcap/serialization.py", "github:foxglove/mcap/python/mcap/mcap/serialization.py"],
        ["mcap/stream_reader.py", "github:foxglove/mcap/python/mcap/mcap/stream_reader.py"],
        ["mcap/summary.py", "github:foxglove/mcap/python/mcap/mcap/summary.py"],
        ["mcap/well_known.py", "github:foxglove/mcap/python/mcap/mcap/well_known.py"],
        ["mcap/writer.py", "github:foxglove/mcap/python/mcap/mcap/writer.py"]
    ],
    "deps": [
        ["abc", "latest"],
        ["collections-defaultdict", "latest"]
    ],
    "version": "1.1.0"
}
```

Next we encounter numerous missing and incomplete packages in MicroPython. MicroPython targets compatibility with Python 3.4, but there are [many documented differences from CPython](https://docs.micropython.org/en/latest/genrst/index.html). Some that affect MCAP, which uses modern Python features, include:

1. Typing problems: There is no `typing` module. I created a dummy one:

    ```python
    class Dict: pass
    class List: pass
    class IO: pass
    # etc. etc.
    ```

2. The `io` module has `IOBase` but not `RawIOBase`, `BufferedReader`, or `BufferedWriter`.

3. The `abc` module does not have `ABC`.

4. The `OrderedDict` comes from the `ucollections` module for some reason.

5. The modules `dataclasses` and `enum` are not implemented.

6. The modules `lz4` and `zstandard` are not available/tested under MicroPython.

Finally, some minor syntax issues:

1. In a few places, f-strings are implicitly concatenated (`f"hello " f"world"`). This must be explicit in MicroPython.

2. It didn't like the subscripting in the type definition of `QueueItem` in _message_queue.py.

**Comment by defunctzombie** [comment]
Probably fair to say that we'd entertain reviewing some PRs around this effort but are unlikely to prioritize this ourselves soon.

**Comment by rgov** [comment]
Here's my stripped down MicroPython port: https://github.com/rgov/micropython-mcap. It runs on my macOS machine and produces a file that, on brief inspection, Foxglove Studio can open without complaint.

The commit history describes the hacks I needed to apply. It's too invasive to suggest merging it into the main source tree, but maybe [this one](https://github.com/rgov/micropython-mcap/commit/c8780526993e13c6f0204220332c5eeab829d297) is worth taking into the mainline?

I haven't tested it on a real microcontroller yet, nor will I get to it for a little while.
